docs/architecture/decisions/0014-dynamic-datasource-registration.md

```diff
@@ -0,0 +1,106 @@
+
+## Design Considerations
+
+Currently Data Sources project is consumed by GSF library using Dependency Injection. Each GSF library handlers have references to IForecastDataSource and/or IEmissionsDataSource Data Source interfaces.
+To modify this interaction and to make it more dynamic, these are things that required to be considered:
+
+- Public interfaces & Data Records
+
+    Data Sources interfaces and Data Records are internal and available only to certain projects (i.e GSF.CarbonAware). This would require to be changed so consumers can dynamically register those and consume them.
+
+    Effort Level: **Medium**
+
+- Packaging
+
+    Current Data Source project is not allowed to be packaged (nuget package), which would require how this is going to be done in terms of, what the package contains, versioning and where to publish it.
+
+    Effort Level: **Medium**
+
+- Load, Register and Instantiate
+
+    Using technics like **Reflection** and **Assembly Discovery**, it would be possible to load Data Sources assemblies and instantiate classes that implements the interfaces that are available.
+    As an example, this could be done via a Data Source class loader
+
+    ```c#
+    static Assembly[] GetDataSourceAssemblies()
+    {
+            var assemblies = Directory.GetFiles(AppDomain.CurrentDomain.BaseDirectory, "*.dll")
+    .Where(x => x.Contains("CarbonAware.DataSource"))
+    .Select(x => Assembly.Load(AssemblyName.GetAssemblyName(x)));
+            return assemblies.ToArray();
+    }
+    ```
+
+    Then the consumer could be using Reflection to instantiate the classes interested:
+
+    ```c#
+    Type providerType = DataSourceAssembly.GetType("CarbonAware.DataSource.ProviderA");
+
+    var theForecast = Activator.CreateInstance(providerType) as IForecastDataSource;
+    theForecast.GetCarbonIntensityForecastAsync(…) // Invoke the interface method.
+    ```
+
+    This responsibility should be part of a new GSF library subsystem.
+
+    Effort Level: **Medium**
+
+- Data Source manifest
+
+    Configuration information would be required to be part of the Data Source package, so the GSF handlers can interact with it. Properties like where to locate the assembly, what assembly to load, what classes to interact with (i.e., Builders/Factories) therefore Emissions and Forecast data can be retrieved from GSF handlers. Designing this manifest would help to drive the implementation of the other items.
+
+    Effort Level: **Large**
+
+- GSF Enhancements
+
+    Given the fact the current registration is done using D.I., GSF library would require to be changed and enhanced to accommodate discovery, how to load the assemblies that are available and that implement Data Sources interfaces. Also understand the configuration that comes from the manifest, in such a way that all the required properties are available.
```

Comment:
This is a small nit: I would say to spell out Dependency Injection  for more clarity instead of using the abbreviation.